### DISCUSSION TITLE
Changing monochrome icons colour on macOS

### DISCUSSION THREAD

**Discussion by NeighNeighNeigh** [ImageMagick/ImageMagick#5610] [Help]
Version: ImageMagick 7.1.0-49 Q16-HDRI aarch64 20471   

Would anyone have the interest and capacity to help generating a MacOS shell script that uses ImageMagick to change the colour of monochrome icons (.png files)?

# _What_ is this thing?

The script should run along these lines:
1) Recursively search a directory.
2) Identifying all images that only contain one colour (+ alpha). 
3) Replace that one colour with a new value (leaving the alpha as is).
4) Save over the image file.

# …and _why_ is this thing?

I've put together [a workflow](https://github.com/NeighNeighNeigh/Alfred_IconGenerator) for macOS app Alfred that facilitates using symbols (from Apple's SF Symbols library) as icons within Alfred's workflows. This new script would allow changing the colour of all these icons, after the fact - and this would be really handy!

**Comment by fmw42** [comment]
Here is an example of an Imagemagick command.

Icon:

![icon](https://user-images.githubusercontent.com/13458218/193393595-c11c1229-400d-477a-8341-e24a224d49d4.png)

`magick icon.png \( +clone -fill red -colorize 100 \) -colorspace sRGB  -compose multiply -composite icon_red.png`

![icon_red](https://user-images.githubusercontent.com/13458218/193393626-72799b37-20be-4f03-969b-57843183b7d6.png)

**Reply by NeighNeighNeigh** [reply]
Hi Fred, thanks so much for the reply, but this isn't really what I'm after. Your contribution is only addressing part 3 of the requirement, but even then it’s not quite handling this step correctly anyway. For step 3 I need to replace the RGB channels to create a new solid fill colour while retaining the original alpha - I believe [this approach](https://stackoverflow.com/a/55658717) looks promising, but I am unsure how to handle implementing this along with the other requirements into one comprehensive script.

**Reply by NeighNeighNeigh** [reply]
I worry my use of the term monochrome may have been misleading - possibly suggesting the source images are black and white. What I mean is that the source images contain only single solid RGB value. Here is a visual explainer in case more clarity is helpful.
![SCR-20221001-njq-2](https://user-images.githubusercontent.com/4795315/193397619-3472c177-d9d2-4b46-9aee-a37f1c262e1f.jpeg)
Note: the 'Before' and 'After' being shown as separate directories is for illustrative purposes only. In reality they would be the same folder, with the same file names / directory names, just the file contents are updated.

**Comment by fmw42** [comment]
![circle_icon](https://user-images.githubusercontent.com/13458218/193418874-5eab6d25-fe57-459b-9763-c2f1a18ec4e8.png)

`magick circle_icon.png -colorspace gray \( +clone -fill red -colorize 100 \) -colorspace sRGB  -compose multiply -composite circle_icon_red.png`

![circle_icon_red](https://user-images.githubusercontent.com/13458218/193418884-2597e526-2103-4f43-8c40-1309991694cc.png)

**Reply by NeighNeighNeigh** [reply]
Again thanks for your input. That red looks very dark, I assume the updated script is first changing the blue circle to greyscale and then multiplying a red fill over that? If so, that explains the darkness (and isn’t going to work here). But this is again really only an attempt to handle step 3 of the problem- where the real meat of the problem is broader and outlined in the other steps.

**Comment by snibgo** [comment]
Detecting whether an image has the same colour in all non-fully-transparent pixels can be done like this. Windows BAT syntax; adjust for bash.
```
magick ^
  %SRC% ^
  -background Black -alpha Background ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  ( +clone ^
    -scale 1x1 ^
    -alpha off ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  ( -clone 1,0 ^
    -layers Flatten ^
  ) ^
  -delete 0 ^
  -compose Difference -composite ^
  -format "IsOne=%%[fx:maxima>0?0:1]\n" ^
  info:
```
This outputs "IsOne=1" if there is only one colour; otherwise it outputs "IsOne=0".

There may be simpler methods.

A simple method to change that colour to another colour such as "red" is:

```
magick circle_icon.png -fill Red -colorize 100 x.png
```
![x](https://user-images.githubusercontent.com/33812368/193420299-443b4a92-827e-4756-89d0-3bdc5d8a4535.png)

**Comment by fmw42** [comment]
We can only help with the Image processing part. You will have to deal with the directories and looping using whatever tools you need (bash, AppleScript, etc).

With regard to step 2, there can be several issues.

1) the image may be antialiased around the boundary of the opaque part. This will cause many colors, but look like one, plus transparency.

2) there can be any number of colors under the transparent parts and it would look like one color plus transparency. Typically you have either the same color under transparency or black. In those cases and no anti-aliasing, you can check for 1 opaque color and one transparent color, so two unique colors.  But there can be anything under the transparency causing the number of unique colors to be more than 2.

How would you want to handle those cases?

**Comment by fmw42** [comment]
Here is another way to do step 3)

![circle_icon](https://user-images.githubusercontent.com/13458218/193422390-125b6212-cbcc-4bf4-a51d-d7769083e64c.png)

`magick circle_icon.png -alpha extract +write mpr:alpha +level-colors red,red mpr:alpha -compose copy_opacity -composite circle_icon_red.png`

![circle_icon_red](https://user-images.githubusercontent.com/13458218/193422394-3d487028-d31b-428d-b35f-27535ed4eb7c.png)

**Comment by snibgo** [comment]
@NeighNeighNeigh: Please show sample files. The actual files, not screenshots of some program that is displaying those files.

**Comment by NeighNeighNeigh** [comment]
Hey guys, thanks again for your time and attention with this.
@snibgo I have attached sample images to this post.¹

Please let me provide an outline of the workflow linked in my first post:

1) I use ImageMagick's caption feature to generate a .png file containing a single text character (a symbol) on a transparent bg ² 
2) I bring these .png's into Alfred.app to use as icons. In this process Alfred.app unavoidably strips any metadata and renames the .png's with a random UUID (see sample files attached to this post).
3) Later I want to change the colour of all these icons. 

So I need a way of identifying which images need adjusting (leaving others untouched) – without the advent of filename / metadata.

Here's something I was thinking about overnight:

I wonder if there is a way to generate some tag-like identifier _within_ the pixels of the image, hidden by the alpha channel. For example, maybe the first 5 pixels are some specific sequence of RGB values, which would be hidden behind the alpha channel. Then when searching through the images, ImageMagick could look for that _'signature'_ when identifying which images to modify. Would that be possible?

This approach (or something similar) would be _ideal_ as we would confidently know that only the _'signed'_ icons would be affected. 

Anyway, I thank you again for spending some time with me on this.

① [Samples.zip](https://github.com/ImageMagick/ImageMagick/files/9691790/Samples.zip)

② ```convert -background none -font /Library/Fonts/SF-Pro-Display-Bold.otf -fill $colour -gravity "center" -size "256x256"  caption:"$query" +repage -depth 8 "$output"```

**Comment by fmw42** [comment]
![airplane](https://user-images.githubusercontent.com/13458218/193438157-2065589f-3721-4858-869f-a44bcc23ad58.png)

`convert airplane.png -alpha extract +write mpr:alpha +level-colors red,red mpr:alpha -compose copy_opacity -composite airplane_red.png`

![airplane_red](https://user-images.githubusercontent.com/13458218/193438160-52ad0743-25cf-45c8-8e91-f58b26119372.png)

See stegano. It distributes the pixels of some message image over the entire image.  I do not know of anything quite like what you request in some localized place that would not be seen if one removes the alpha channel.  You can simple overlay one small image onto any part of the image under the alpha channel and have it not be seen. But if the alpha channel is removed it would be seen.

**Reply by NeighNeighNeigh** [reply]
It wouldn’t matter at all if the ‘signature pixels’ are visible when removing the alpha channel. So maybe this approach has legs. That is if there’s a way to use ImageMagick to ‘find’ files based on the contents of that small portion of the image?

**Comment by snibgo** [comment]
> 2. Identifying all images that only contain one colour (+ alpha).

> What I mean is that the source images contain only single solid RGB value. 

Those statements are not precise. I thought they meant that all pixels that are not fully transparent are the same colour, and I showed a "detection" command for this. But your examples show this is not the case, so my command doesn't do what you need. Perhaps all pixels that are fully-opaque are the same colour. Or perhaps something else.

But the simple conversion works: we can change all pixels to red, retaining alpha, like this:
```
magick 40DB65DE-31C2-4074-ADBB-F0616D2EA04B.png -fill Red -colorize 100 x.png
```
![x](https://user-images.githubusercontent.com/33812368/193459739-21e7d0ae-01cb-49b8-9ad9-cbc0f6d1e51f.png)

**Reply by NeighNeighNeigh** [reply]
Apologies for not having provided precision in my terminology and understanding. I made some incorrect assumptions and didn't realise that underneath the alpha my images were in fact not one solid colour. 

I have a feeling your suggested approach "all pixels that are fully-opaque are the same colour" might be a good way to tackle this… would you have an idea for how this might be implemented?

**Reply by fmw42** [reply]
Either way, your image is either the same color or black under the transparency. So you would mostly have 2 colors in the histogram, some opaque RGBA where A=1 or RGBA were A=0.  The RGB part might be the same in both regions or it might be some color or black.  If all your images have been processed by Alfred and RGB turned black, then that ensures that you have 2 colors, provided it is transparent.  (You could have 2 opaque colors and totally opaque alpha, but I suspect you are eliminating that possibility).

**Reply by NeighNeighNeigh** [reply]
Unfortunately, precisely as you flagged earlier, there is some fringing around the edges where other values are being introduced in the RBG channels. Here I have used ImageMagick to remove the alpha channel and I can see clearly some slight differences in values. 

Zoomed in screenshot:
![SCR-20221004-i11](https://user-images.githubusercontent.com/4795315/193718251-4a58c250-ac0e-4c5e-bb0a-94c0b32c0763.png)

Source:
![testnoalpha](https://user-images.githubusercontent.com/4795315/193718283-9492a5e8-a4ba-4edc-afce-196e848fe4c1.png)

**Reply by NeighNeighNeigh** [reply]
this is where some "fuzziness" can be helpful perhaps... since there is _almost only_ 2 colours in the histogram... can some kind of threshold be factored in maybe?

**Reply by fmw42** [reply]
Are these odd colors in the original or from Alfred processing?

If you want to know if the opaque region is one color within some tolerance you can do the following.  

I note that is the (opaque) color is solid, the standard deviation will be zero. 

`magick airplane.png -colorspace gray -channel-fx "alpha=>read-mask" -channel rgb -format "%[fx:standard_deviation]\n" info:
`

0.000185355

So this shows that you have a few pixels that are off (as you pointed out).

`magick airplane.png -colorspace gray +write t1.png -channel-fx "alpha=>read-mask" -channel rgb -format "%[fx:(standard_deviation<=T)?1:0]\n" info:
`

If T=0.0002, then the result is 1
If T=0.0001 then the result is 0

Since the std is 0.000185355, a threshold of T=0.0002 works to say that the image has nearly constant color in the opaque region.

where T is in the range 0.0 to 1.0

Pick your threshold according to tests of a number of your images.

**Reply by NeighNeighNeigh** [reply]
Awesome thank you fmw42, this looks like a great lead. With this in hand I now need to return to the drawing board, and work out the non-Imagemagick aspects of the problem. Thanks for you help with this so far, leave it with me for now

**Comment by NeighNeighNeigh** [comment]
Hey guys,

I got very excited because I pretty much got this 'pixel signature' working. I learned that I can use ImageMagick's compare function to evaluate the similarity between 2 inputs. So if I included a tiny line of signature pixels, hidden behind the alpha in my source image - when searching for 'signed' images, I could first delete the alpha, then crop to the signature region, and finally compare it to another image containing the same signature. 

This was all working great…

Until I realised that Alfred.app was re-writing the RGB values of fully transparent pixels to black, thus removing my little signatures. 🤦‍♂️

So I'm back to square one. I need to try and evaluate the image _even though_ there is potential for fringe values around the edges and black in areas of zero alpha. A fools errand?

**Reply by fmw42** [reply]
Would Alfred overwrite your RGB image if it was already black with your small signature?  Do you have any control of what Alfred does when processing?

**Reply by fmw42** [reply]
Can you not just write to either the comment or label in the header with Imagemagick -label for example?  Have you tried using stegano in the alpha channel. I suspect that your message would not be visible and you could reconstitute it later.

**Reply by NeighNeighNeigh** [reply]
I started out by attempting to use -label, and sadly it was no good. Alfred seems to be generating new files and omitting such information. I have no control over how Alfred handles this process… and yeah Alfred is replacing all RGB values with black that are opacity zero.

**Reply by NeighNeighNeigh** [reply]
Im not sure stegano will work, since Alfred is messing with the values

**Comment by snibgo** [comment]
> I have a feeling your suggested approach "all pixels that are fully-opaque are the same colour" might be a good way to tackle this… would you have an idea for how this might be implemented?

We can modify the test I showed above, by first changing alpha so that all pixels that are not fully opaque (A=1) are made fully transparent (A=0). Windows BAT syntax:
```
magick ^
  %SRC% ^
  -channel A -fx "u<1?0:1" +channel ^
  -background Black -alpha Background ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  ( +clone ^
    -scale 1x1 ^
    -alpha off ^
    -scale "%%[MYSIZE]^!" ^
  ) ^
  ( -clone 1,0 ^
    -layers Flatten ^
  ) ^
  -delete 0 ^
  -compose Difference -composite ^
  -format "IsOpaqOne=%%[fx:maxima>0?0:1]\n" ^
  info:
```
This returns IsOpaqOne=1 for your images.

**Reply by NeighNeighNeigh** [reply]
Okay this is great thanks Snibgo.. I first need to work out how to translate from bat to zsh, but Ill try tackle that myself
